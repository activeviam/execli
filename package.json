{
  "name": "execli",
  "description": "Generate task-oriented CLIs declaratively",
  "version": "3.0.0-beta.0",
  "keywords": [
    "cli",
    "task",
    "exec",
    "listr",
    "yargs"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/activeviam/execli"
  },
  "engines": {
    "node": "^14.13.1 || >=16.0.0"
  },
  "bugs": {
    "url": "https://github.com/activeviam/execli/issues"
  },
  "files": [
    "lib"
  ],
  "bin": "lib/bin.js",
  "exports": "./lib/index.js",
  "scripts": {
    "build": "tsc --build tsconfig.json",
    "check-prettier": "yarn run prettier --check",
    "eslint": "eslint --ignore-path .gitignore --max-warnings 0 \"./**/*.{js,ts}\"",
    "format-prettier": "yarn run prettier --write",
    "prettier": "prettier --ignore-path .gitignore \"./**/*.{cjs,js,json,md,ts,yml}\"",
    "pretest": "tsc --build tsconfig.test.json",
    "test": "jest"
  },
  "dependencies": {
    "@types/fs-extra": "^9.0.12",
    "@vercel/ncc": "^0.29.1",
    "execa": "^5.1.1",
    "listr2": "^3.11.0",
    "loud-rejection": "^2.2.0",
    "slugify": "^1.6.0",
    "yargs": "^17.1.1"
  },
  "devDependencies": {
    "@babel/core": "^7.15.0",
    "@babel/preset-env": "^7.15.0",
    "@babel/preset-typescript": "^7.15.0",
    "@types/jest": "^27.0.1",
    "@types/yargs": "^17.0.2",
    "@typescript-eslint/eslint-plugin": "^4.29.2",
    "@typescript-eslint/parser": "^4.29.2",
    "babel-jest": "^27.0.6",
    "eslint": "^7.32.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-config-xo": "^0.38.0",
    "eslint-config-xo-typescript": "^0.44.0",
    "eslint-import-resolver-typescript": "^2.4.0",
    "eslint-plugin-import": "^2.24.0",
    "eslint-plugin-sort-destructure-keys": "^1.3.5",
    "eslint-plugin-typescript-sort-keys": "^1.8.0",
    "eslint-plugin-unicorn": "^35.0.0",
    "jest": "^27.0.6",
    "pkg-dir": "^5.0.0",
    "prettier": "^2.3.2",
    "tempy": "^1.0.1",
    "typescript": "^4.3.5",
    "yarn-deduplicate": "^3.1.0"
  },
  "type": "module"
}
