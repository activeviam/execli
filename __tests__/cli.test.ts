import path from "path";
import execa from "execa";
import pkgDir from "pkg-dir";
import tempy from "tempy";
import { bin } from "../package.json";

const binPath = path.join(pkgDir.sync() as string, bin);
const commandFilePath = path.join(__dirname, "command.json");

test("compile", async () => {
  // Use a temporary file path outside the package to make sure all the dependencies
  // are correctly bundled and are not required through the package's node_modules.
  const targetPath = tempy.file({ extension: "js" });
  await execa("node", [binPath, "compile", commandFilePath, targetPath]);
  const { stdout } = await execa("node", [
    targetPath,
    "test-command",
    "--debug",
  ]);
  expect(stdout).toContain("Print current directory");
  expect(stdout).toContain(process.cwd());
}, 10000);

describe("run", () => {
  const runArguments = [binPath, "run", commandFilePath];
  const runTestCommandArguments = [...runArguments, "--", "test-command"];

  test.each([
    [
      "command required",
      runArguments,
      {
        failed: true,
        outputs: ["Not enough non-option arguments: got 0, need at least 1"],
      },
    ],
    [
      "strict (existing-option)",
      [...runTestCommandArguments, "--flag"],
      {
        outputs: ["Print current directory"],
      },
    ],
    [
      "strict (unexisting-option)",
      [...runTestCommandArguments, "--unexisting-option"],
      {
        failed: true,
        outputs: ["Unknown arguments: unexisting-option"],
      },
    ],
    [
      "debug",
      [...runTestCommandArguments, "--debug"],
      {
        outputs: ["Print current directory", process.cwd()],
      },
    ],
    [
      "only (exact)",
      [...runTestCommandArguments, "--only", "Print current directory"],
      {
        outputs: [
          "Print current directory [completed]",
          "List content of directory [skipped]",
          "List current processes [skipped]",
        ],
      },
    ],
    [
      "only (slug)",
      [...runTestCommandArguments, "--only", "print-current-directory"],
      {
        outputs: [
          "Print current directory [completed]",
          "List content of directory [skipped]",
          "List current processes [skipped]",
        ],
      },
    ],
    [
      "only (no-match)",
      [...runTestCommandArguments, "--only", "PrInt-CuRRenT-DirEctoRy"],
      {
        failed: true,
        outputs: ["Invalid values"],
      },
    ],
    [
      "skip (exact)",
      [...runTestCommandArguments, "--skip", "Print current directory"],
      {
        outputs: [
          "Print current directory [skipped]",
          "List content of directory [completed]",
          "List current processes [completed]",
        ],
      },
    ],
    [
      "skip (slug)",
      [...runTestCommandArguments, "--skip", "print-current-directory"],
      {
        outputs: [
          "Print current directory [skipped]",
          "List content of directory [completed]",
          "List current processes [completed]",
        ],
      },
    ],
    [
      "skip (no-match)",
      [...runTestCommandArguments, "--skip", "PrInt-CuRRenT-DirEctoRy"],
      {
        failed: true,
        outputs: ["Invalid values"],
      },
    ],
    [
      "tag (existing)",
      [...runTestCommandArguments, "--tag", "a"],
      {
        outputs: [
          "Print current directory [completed]",
          "List content of directory [skipped]",
          "List current processes [skipped]",
        ],
      },
    ],
    [
      "tag (unexisting)",
      [...runTestCommandArguments, "--tag", "d"],
      {
        failed: true,
        outputs: ["Invalid values"],
      },
    ],
  ])(
    "%s %j",
    async (
      _,
      commandArguments: readonly string[],
      {
        failed = false,
        outputs,
      }: { readonly failed?: boolean; readonly outputs: readonly string[] },
    ) => {
      const result = await execa("node", commandArguments, {
        reject: !failed,
      });
      expect(result.failed).toBe(failed);
      outputs.forEach((output) => {
        expect(failed ? result.stderr : result.stdout).toContain(output);
      });
    },
  );
});
