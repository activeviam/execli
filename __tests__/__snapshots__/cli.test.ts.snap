// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run background command {"commandName":"background"} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Parent task [starting background process]",
    "[DATA] node --eval \\"setTimeout(() => {console.log('f' + 'o' + 'o'); setTimeout(() => {}, 10000)}, 500)\\"",
    "[SUCCESS] Parent task [starting background process]",
    "[STARTED] Parent task [with background process]",
    "[STARTED] Print matched output",
    "[DATA] echo foo",
    "foo",
    "[SUCCESS] Print matched output",
    "[SUCCESS] Parent task [with background process]",
    "[STARTED] Parent task [stopping background process]",
    "[SUCCESS] Parent task [stopping background process]",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run command required {"failing":true} 1`] = `
Object {
  "stderr": Array [
    "Not enough non-option arguments: got 0, need at least 1",
  ],
}
`;

exports[`run command task not matching regexp {"commandName":"error","failing":true,"options":["--from","Run unmatched background command"]} 1`] = `
Object {
  "stderr": Array [
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "Error: Background process exited before matching the regexp. Command was: ls",
    "    at ExecError.toDetailedError (file://EXECLI_DIRECTORY/lib/exec.js:22:16)",
    "    at runTask (file://EXECLI_DIRECTORY/lib/tasks.js:385:25)",
    "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "    at async Object.handler (file://EXECLI_DIRECTORY/lib/commands.js:15:9)",
  ],
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Print current directory",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Nest task",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Run broken copy command",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Run broken background command",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Run unmatched background command",
    "[STARTED] Run unmatched background command [starting background process]",
    "[DATA] ls",
  ],
}
`;

exports[`run debug {"commandName":"simple","options":["--debug"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run debug forced in non interactive terminals {"commandName":"simple","failing":true,"options":["--no-debug"]} 1`] = `
Object {
  "stderr": Array [
    "Cannot opt-out of debug mode in non interactive terminals.",
  ],
}
`;

exports[`run dryRun {"commandName":"simple","options":["--dryRun"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[TITLE] Echo start ($ echo start)",
    "[SKIPPED] Skipped by --dryRun option",
    "[STARTED] Echo middle",
    "[TITLE] Echo middle ($ echo mystery)",
    "[SKIPPED] Skipped by --dryRun option",
    "[STARTED] Echo end",
    "[TITLE] Echo end ($ echo end)",
    "[SKIPPED] Skipped by --dryRun option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run failed command task {"commandName":"error","failing":true,"options":["--from","Run broken copy command"]} 1`] = `
Object {
  "stderr": Array [
    "cp: unexisting-source: No such file or directory",
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "Error: Command failed with exit code 1: cp unexisting-source target",
    "    at ExecError.toDetailedError (file://EXECLI_DIRECTORY/lib/exec.js:22:16)",
    "    at runTask (file://EXECLI_DIRECTORY/lib/tasks.js:385:25)",
    "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "    at async Object.handler (file://EXECLI_DIRECTORY/lib/commands.js:15:9)",
  ],
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Print current directory",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Nest task",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Run broken copy command",
    "[DATA] cp unexisting-source target",
  ],
}
`;

exports[`run failed regular task {"commandName":"error","failing":true} 1`] = `
Object {
  "stderr": Array [
    "[FAILED] Something went wrong.",
    "[FAILED] Something went wrong.",
    "[FAILED] Something went wrong.",
    "Error: Something went wrong.",
    "    at run (file://EXECLI_DIRECTORY/lib/__test_resources__/commands/error.js:39:35)",
    "    at runRegularTask (file://EXECLI_DIRECTORY/lib/tasks.js:181:26)",
    "    at Task.task (file://EXECLI_DIRECTORY/lib/tasks.js:212:15)",
    "    at Task.run (EXECLI_DIRECTORY/node_modules/listr2/dist/lib/task.js:213:45)",
    "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
  ],
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Print current directory",
    "[DATA] pwd",
    "EXECLI_DIRECTORY",
    "[SUCCESS] Print current directory",
    "[STARTED] Nest task",
    "[STARTED] Throw exception",
  ],
}
`;

exports[`run from {"commandName":"simple","options":["--from","Echo middle"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --from option",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run only (exact title) {"commandName":"simple","options":["--only","Echo start"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --only option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --only option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run only (neither exact title nor slug) {"commandName":"simple","failing":true,"options":["--only","EcHo-StArt"]} 1`] = `
Object {
  "stderr": Array [
    "Invalid values:",
    "  Argument: only, Given: \\"EcHo-StArt\\", Choices: \\"Echo end\\", \\"Echo middle\\", \\"Echo start\\", \\"Parent task\\"",
  ],
}
`;

exports[`run only (slug) {"commandName":"simple","options":["--only","echo-start"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --only option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --only option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run option (existing) {"commandName":"simple","options":["--flag"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run option (unexisting) {"commandName":"simple","failing":true,"options":["--unexisting-option"]} 1`] = `
Object {
  "stderr": Array [
    "Unknown arguments: unexisting-option, unexistingOption",
  ],
}
`;

exports[`run rollback {"commandName":"rollback","failing":true} 1`] = `
Object {
  "stderr": Array [
    "cp: unexisting-source: No such file or directory",
    "[FAILED] Command failed",
    "[FAILED] Command failed",
    "[ROLLBACK] Parent task",
    "Error: Parent task",
    "    at Task.run (EXECLI_DIRECTORY/node_modules/listr2/dist/lib/task.js:257:27)",
    "    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "    at async EXECLI_DIRECTORY/node_modules/p-map/index.js:57:22",
  ],
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Run broken copy command",
    "[DATA] cp unexisting-source target",
    "[DATA] echo rollback",
    "rollback",
  ],
}
`;

exports[`run skip (exact title) {"commandName":"simple","options":["--skip","Echo start"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --skip option",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run skip (neither exact title nor slug) {"commandName":"simple","failing":true,"options":["--skip","EcHo-StArt"]} 1`] = `
Object {
  "stderr": Array [
    "Invalid values:",
    "  Argument: skip, Given: \\"EcHo-StArt\\", Choices: \\"Echo end\\", \\"Echo middle\\", \\"Echo start\\", \\"Parent task\\"",
  ],
}
`;

exports[`run skip (slug) {"commandName":"simple","options":["--skip","echo-start"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[SKIPPED] Skipped by --skip option",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[DATA] echo end",
    "end",
    "[SUCCESS] Echo end",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run tag (existing) {"commandName":"simple","options":["--tag","a"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[SKIPPED] Skipped by --tag option",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --tag option",
    "[SUCCESS] Parent task",
  ],
}
`;

exports[`run tag (unexisting) {"commandName":"simple","failing":true,"options":["--tag","d"]} 1`] = `
Object {
  "stderr": Array [
    "Invalid values:",
    "  Argument: tag, Given: \\"d\\", Choices: \\"a\\", \\"b\\", \\"c\\"",
  ],
}
`;

exports[`run until {"commandName":"simple","options":["--until","Echo middle"]} 1`] = `
Object {
  "stdout": Array [
    "[STARTED] Parent task",
    "[STARTED] Echo start",
    "[DATA] echo start",
    "start",
    "[SUCCESS] Echo start",
    "[STARTED] Echo middle",
    "[DATA] echo mystery",
    "mystery",
    "[SUCCESS] Echo middle",
    "[STARTED] Echo end",
    "[SKIPPED] Skipped by --until option",
    "[SUCCESS] Parent task",
  ],
}
`;
